pulsar-client/src/main/java/com/yahoo/pulsar/client/impl/PartitionedConsumerImpl.java

```diff
@@ -42,19 +44,28 @@
 
 public class PartitionedConsumerImpl extends ConsumerBase {
 
-    private List<ConsumerImpl> consumers;
-    private int numPartitions;
-    private final ExecutorService internalListenerExecutor;
+    private final List<ConsumerImpl> consumers;
+
+    // Queue of partition consumers on which we have stopped calling receiveAsync() because the
+    // shared incoming queue was full
+    private final ConcurrentLinkedQueue<ConsumerImpl> pausedConsumers;
+
+    // Threshold for the shared queue. When the size of the shared queue goes below the threshold, we are going to
+    // resume receiving from the paused consumer partitions
+    private final int sharedQueueResumeThreshold;
+
+    private final int numPartitions;
     private final ReadWriteLock lock = new ReentrantReadWriteLock();
     private final ConsumerStats stats;
 
     PartitionedConsumerImpl(PulsarClientImpl client, String topic, String subscription, ConsumerConfiguration conf,
             int numPartitions, ExecutorService listenerExecutor, CompletableFuture<Consumer> subscribeFuture) {
-        super(client, topic, subscription, conf, listenerExecutor, subscribeFuture, false /* use fixed size queue */ );
+        super(client, topic, subscription, conf, listenerExecutor, subscribeFuture);
         this.consumers = Lists.newArrayListWithCapacity(numPartitions);
+        this.pausedConsumers = new ConcurrentLinkedQueue<>();
+        this.sharedQueueResumeThreshold = maxReceiverQueueSize / 2;
```

Comment:
having the threshold too tight would make flip-flopping easy: receiving a single message would make pausing the consumer and then when the application process it, resume the consumer. Using half-the queue size is to reduce that contention, and at the same time avoids the queue to get empty and starve the application